Reading_and_Writing_Data/Creating_Efficient_GeoParquet_Files.ipynb

```diff
@@ -0,0 +1,660 @@
+    "- Uses wildcard characters to recursively grab all files regardless of year or quarter.\n",
+    "- Reads the data into a Spark DataFrame using Sedona's reader.\n",
+    "\n",
+    "---\n",
+    "\n",
+    "\n",
+    "```python\n",
+    "    .withColumn(\"file_path\", input_file_name()) \\\n",
+    "```\n",
+    "**What this does:**\n",
+    "- Adds a new column called `file_path` containing the S3 path where each row came from.\n",
+    "- This column is used to extract temporal metadata.\n",
+    "\n",
+    "---\n",
+    "\n",
+    "\n",
+    "```python\n",
+    "    .withColumn(\"year\", regexp_extract(\"file_path\", r\"year=(\\\\d+)\", 1)) \\\n",
+    "```\n",
+    "**What this does:**\n",
+    "- Extracts the four-digit `year` from the file path using a regex pattern and stores it in a new column.\n",
+    "\n",
+    "---\n",
+    "\n",
+    "```python\n",
+    "    .withColumn(\"quarter\", regexp_extract(\"file_path\", r\"quarter=(\\\\d+)\", 1)) \\\n",
+    "```\n",
+    "**What this does:**\n",
+    "- Extracts the `quarter` (1 to 4) from the file path using another regex pattern.\n",
+    "\n",
+    "---\n",
+    "\n",
+    "```python\n",
+    "    .withColumn(\"geometry\", expr(\"ST_GeomFromText(tile)\")) \\\n",
+    "    \n",
+    "```\n",
+    "**What this does:**\n",
+    "- Converts the `tile` column (in WKT format) into a geometry column using Sedona's `ST_GeomFromText()` function.\n",
+    "- This enables spatial operations on the geometries.\n",
+    "\n",
+    "---\n",
+    "\n",
+    "```python\n",
+    "    .withColumn(\"bbox\", expr(\"struct(st_xmin(ST_GeomFromText(tile)) as xmin, st_ymin(ST_GeomFromText(tile)) as ymin, st_xmax(ST_GeomFromText(tile)) as xmax, st_ymax(ST_GeomFromText(tile)) as ymax) as bbox\")) \\\n",
+    "```\n",
+    "**What this does:**\n",
+    "- Constructs a bounding box for each geometry by extracting the min and max x and y coordinates.\n",
+    "- Creates a struct (`bbox`) with fields `xmin`, `ymin`, `xmax`, and `ymax` for spatial indexing or filtering.\n",
+    "\n",
+    "---\n",
+    "\n",
+    "\n",
+    "```python\n",
+    "    .withColumn(\"geohash\", expr(\"ST_GeoHash(ST_GeomFromText(tile), 10)\")) \\\n",
+    "```\n",
+    "**What this does:**\n",
+    "- Generates a geohash with precision 10 for each geometry using Sedona.\n",
+    "- Geohashing encodes spatial location into alphanumeric strings and is useful for spatial partitioning or clustering.\n",
+    "\n",
+    "---\n",
+    "\n",
+    "\n",
+    "```python\n",
+    "    .selectExpr(\"*\", ''' \"mobile\" as type''') \\\n",
+    "```\n",
+    "**What this does:**\n",
+    "- Adds a new column called `type` and sets its value to \"mobile\" for all rows.\n",
+    "- Helps distinguish this dataset from others, such as fixed broadband data.\n",
+    "\n",
+    "---\n",
+    "\n",
+    "```python\n",
+    "    .orderBy(expr(\"ST_GeoHash(ST_GeomFromText(tile), 6)\")) \\\n",
+    "```\n",
+    "**What this does:**\n",
```

Comment:
this "what this does" could be omitted since code is self explanatory. repetitive "what this does" could be removed and markdown explanations refactored to more narrative style.